backend/inspirehep/disambiguation/tasks.py

```diff
@@ -70,26 +72,97 @@ def disambiguate_signatures(self, clusters):
     db.session.commit()
 
 
+def match_literature_author(author_data, matcher_config):
+    matched_records = (
+        matched_record["inner_hits"]["authors"]["hits"]["hits"]
+        for matched_record in match(author_data, matcher_config)
+    )
+    matched_authors_refs = []
+    for record_authors in matched_records:
+        matched_authors_refs.extend(
+            author["_source"]["record"]["$ref"]
+            for author in record_authors
+            if "record" in author["_source"]
+        )
+    if len(set(matched_authors_refs)) == 1:
+        author_reference_url = matched_authors_refs[0]
+        return {"$ref": author_reference_url}
+
+
+def assign_reference_to_author(author, updated_authors, reference):
+    author["record"] = reference
+    updated_authors.append(reference["$ref"].split("/")[-1])
+
+
+def match_literature_author_with_configs(author, updated_authors, record):
+    configs = [
+        current_app.config["AUTHOR_MATCHER_NAME_CONFIG"],
+        current_app.config["AUTHOR_MATCHER_NAME_INITIALS_CONFIG"],
+        current_app.config["AUTHOR_MATCHER_COLLABORATION_CONFIG"],
+        current_app.config["AUTHOR_MATCHER_AFFILIATION_CONFIG"],
+    ]
+
+    parsed_name = ParsedName.loads(author.get("full_name"))
+    author_matcher_data = {
+        "first_name": parsed_name.first,
+        "last_name": parsed_name.last,
+        "full_name": author.get("full_name"),
+        "collaborations": get_value(record, "collaborations.value", []),
+        "affiliations": get_value(author, "affiliations.value", []),
+    }
+
+    for config in configs:
+        matched_recid = match_literature_author(author_matcher_data, config)
+        if matched_recid:
+            assign_reference_to_author(author, updated_authors, matched_recid)
+            return author
+
+
 @shared_task(ignore_result=False, bind=True)
 def disambiguate_authors(self, record_uuid):
     record = InspireRecord.get_record(record_uuid)
     if "Literature" not in record["_collections"]:
         return
     authors = record.get_modified_authors()
     updated_authors = []
+
     for author in authors:
         if author.get("curated_relation"):
             continue
-        matched_authors = [
+        matched_authors = {
             matched_record["_source"]["control_number"]
             for matched_record in match(
                 author, current_app.config["AUTHOR_MATCHER_EXACT_CONFIG"]
             )
-        ]
-        if len(set(matched_authors)) == 1:
-            author_control_number = matched_authors[0]
-            author["record"] = get_ref_from_pid("aut", author_control_number)
-            updated_authors.append(str(author_control_number))
+        }
```

Comment:
any reason this couldn't be written as part of `match_literature_author`? it looks like it has the same structure as the other cases (if you retrieve `self` instead of `control_number` from the author record).

Reply:
Yes, I wanted to do it in this way but I am not sure if it would be better. In this case I have to add two conditions in `match_literature_author_config`- first one because I don't get data from `inner_hits` but from `_source` and secondly because I will take reference from `self`, not `record` as in other cases.

Reply:
ok, if it's very different let's keep it separate. Maybe split it out to a different function though to avoid different "levels" of logic in the function?